### PR TITLE
8315856: RISC-V: Use Zacas extension for cmpxchg

### PR DESCRIPTION
**Pull request by luhenry**
---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8315856](https://bugs.openjdk.org/browse/JDK-8315856): RISC-V: Use Zacas extension for cmpxchg (**Enhancement** - P4)

### Reviewers
 * [Robbin Ehn](https://openjdk.org/census#rehn) (@robehn - **Reviewer**)
 * [Fei Yang](https://openjdk.org/census#fyang) (@RealFYang - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk.git pull/16910/head:pull/16910` \
`$ git checkout pull/16910`

Update a local copy of the PR: \
`$ git checkout pull/16910` \
`$ git pull https://git.openjdk.org/jdk.git pull/16910/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 16910`

View PR using the GUI difftool: \
`$ git pr show -t 16910`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/16910.diff">https://git.openjdk.org/jdk/pull/16910.diff</a>

</details>

### Webrev
[Link to Webrev Comment](https://git.openjdk.org/jdk/pull/16910#issuecomment-1834279298)

### CODE REVIEW

File: src/hotspot/cpu/riscv/macroAssembler_riscv.cpp [hunk 1]
```diff
@@ -2624,20 +2624,27 @@ void MacroAssembler::cmpxchgptr(Register oldv, Register newv, Register addr, Reg
...
+    // Fail and exit if it is not what we expect
+    bne(tmp, oldv, nope);
+    // If the store conditional succeeds, tmp will be zero
+    store_conditional(tmp, newv, addr, int64, Assembler::rl);
+    beqz(tmp, succeed);
+    // Retry only when the store conditional failed
+    j(retry_load);
+
+    bind(nope);
+    membar(AnyAny);
```
Comment (robehn) [inline src/hotspot/cpu/riscv/macroAssembler_riscv.cpp:2645]:
```suggestion
    membar(AnyAny);
```

File: src/hotspot/cpu/riscv/macroAssembler_riscv.cpp [hunk 2]
```diff
@@ -2624,20 +2624,27 @@ void MacroAssembler::cmpxchgptr(Register oldv, Register newv, Register addr, Reg
...
+  }
+
```
Comment (robehn) [inline src/hotspot/cpu/riscv/macroAssembler_riscv.cpp:2647]:
```suggestion
    membar(AnyAny);
```

File: src/hotspot/cpu/riscv/macroAssembler_riscv.cpp (hunk 1 above)
Comment (robehn) [inline src/hotspot/cpu/riscv/macroAssembler_riscv.cpp:2645]:
```suggestion
    membar(AnyAny);
```

**Review by robehn — COMMENTED** [review]
I this was what I think we had somewhere :)

File: src/hotspot/cpu/riscv/macroAssembler_riscv.hpp [hunk 3]
```diff
@@ -1063,8 +1063,25 @@ class MacroAssembler: public Assembler {
...
+  void atomic_casal(Register prev, Register newv, Register addr);
+  void atomic_casalw(Register prev, Register newv, Register addr);
+  void atomic_caswu(Register prev, Register newv, Register addr);
+  void atomic_caslwu(Register prev, Register newv, Register addr);
+  void atomic_casalwu(Register prev, Register newv, Register addr);
+
+  void atomic_cas(Register prev, Register newv, Register addr, enum operand_size size,
+              Assembler::Aqrl acquire = Assembler::relaxed, Assembler::Aqrl release = Assembler::relaxed);
+
+  static bool far_branches() {
```
Comment (robehn) [inline src/hotspot/cpu/riscv/macroAssembler_riscv.hpp:1079]: Is this related ?

Comment (luhenry) [reply src/hotspot/cpu/riscv/macroAssembler_riscv.hpp:1079]: Not at all, it's a merge-conflict-fix gone wrong. Let me fix right now.

File: src/hotspot/cpu/riscv/macroAssembler_riscv.cpp [hunk 4]
```diff
@@ -2624,20 +2624,27 @@ void MacroAssembler::cmpxchgptr(Register oldv, Register newv, Register addr, Reg
...
-  beqz(tmp, succeed);
-  // Retry only when the store conditional failed
-  j(retry_load);
-
-  bind(nope);
-  membar(AnyAny);
+  if (UseZacas) {
+    mv(tmp, oldv);
+    atomic_cas(tmp, newv, addr, Assembler::int64, Assembler::aq, Assembler::rl);
+    beq(tmp, oldv, succeed);
```
[the remark this answers is not in the document]
Comment (theRealAph) [reply src/hotspot/cpu/riscv/macroAssembler_riscv.cpp:2738]: That dmb is not present in the AArch64 port because we want a release when the CAS fails, because if it fails nothing was stored, so there is literally nothing for a subsequent load from that address to synchronize with.

It's there because of this re-ordering:
```


	<Access [A]>

	// atomic_op (B)
1:	ldaxr	x0, [B]		// Exclusive load with acquire
	<op(B)>
	stlxr	w1, x0, [B]	// Exclusive store with release
	cbnz	w1, 1b

	<Access [C]>

It doesn't forbid orderings such as

Load [B] -> Load [C] -> Store [A] -> Store [B]
```

[See here](https://mail.openjdk.org/pipermail/aarch64-port-dev/2014-February/000706.html)

The Arm memory model has been strengthened, and this reasoning looks a bit shaky today. At the time we did not know if any of the usages of `cmpxchgptr`required "full barrier" semantics, so we put a full barrier in for safety's sake.

[the remark this answers is not in the document]
Comment (RealFYang) [reply src/hotspot/cpu/riscv/macroAssembler_riscv.cpp:2738]: Wow, thanks for finding that history. It's very helpful for us to understand the existence of this barrier.

File: src/hotspot/cpu/riscv/macroAssembler_riscv.cpp [hunk 5]
```diff
@@ -2844,14 +2853,29 @@ void MacroAssembler::cmpxchg_narrow_value(Register addr, Register expected,
...
+
+    // if old & mask != expected
+    andr(tmp, old, mask);
+    bne(tmp, expected, fail);
+
+    andr(tmp, old, not_mask);
+    orr(tmp, tmp, new_val);
+
+    atomic_cas(old, tmp, aligned_addr, size, acquire, release);
+    bne(tmp, old, retry);
```
Comment (RealFYang) [inline src/hotspot/cpu/riscv/macroAssembler_riscv.cpp:2868]: Does it make sense to use `atomic_cas` here for this case? I don't see the benefit. The input `size` is supposed to be int8 or int16, but the `atomic_cas` can only handle `int64`, `int32` and `uint32`.

File: src/hotspot/cpu/riscv/macroAssembler_riscv.cpp [hunk 6]
```diff
@@ -2891,14 +2915,28 @@ void MacroAssembler::weak_cmpxchg_narrow_value(Register addr, Register expected,
...
-  bnez(tmp, fail);
+    // if old & mask != expected
+    andr(tmp, old, mask);
+    bne(tmp, expected, fail);
+
+    andr(tmp, old, not_mask);
+    orr(tmp, tmp, new_val);
+
+    atomic_cas(tmp, new_val, addr, size, acquire, release);
+    bne(tmp, old, fail);
```
Comment (RealFYang) [inline src/hotspot/cpu/riscv/macroAssembler_riscv.cpp:2929]: Similar question as above here for this case.

**Review by RealFYang — CHANGES_REQUESTED** [review]

File: src/hotspot/cpu/riscv/macroAssembler_riscv.cpp (hunk 4 above)
[the remark this answers is not in the document]
Comment (robehn) [reply src/hotspot/cpu/riscv/macroAssembler_riscv.cpp:2738]: I have seen code, do not find it anymore in code base, which use cmpxchg for Atomic::release_store_fence().
(which is "new");
```
Atomic::store(_x, true);
Atomic::store_release_fence(_one_way_barrier, true);
bool z = Atomic::load(_z);
```
But it is written as:
```
Atomic::store(_x, true);
Atomic::cmpxchg(_one_way_barrier, false, true);
bool z = Atomic::load(_z);
```
Where load Z happens after store X. But without release the store may happens after.

As we say:
"All of the atomic operations that imply a read-modify-write action guarantee a two-way memory barrier across that operation."

I have look over the cmpxchg uses and not releasing 'seems' okay at first glance.
But I think before this line changes we should release on failed CAS.

[the remark this answers is not in the document]
Comment (robehn) [reply src/hotspot/cpu/riscv/macroAssembler_riscv.cpp:2738]: Note I mention the hotspot runtime atomic here because cmpxchgptr is used to manipulate the markword which follows hotspot atomics, not Java volatiles.

File: src/hotspot/cpu/riscv/macroAssembler_riscv.cpp (hunk 5 above)
Comment (luhenry) [reply src/hotspot/cpu/riscv/macroAssembler_riscv.cpp:2868]: It does make sense to use `atomic_cas` here all to avoid `lr/sc`. The manipulation needed for sub-word atomics is the same for both amocas and lr/sc anyway, just a slightly different order.

File: src/hotspot/cpu/riscv/macroAssembler_riscv.cpp (hunk 4 above)
[the remark this answers is not in the document]
Comment (theRealAph) [reply src/hotspot/cpu/riscv/macroAssembler_riscv.cpp:2738]: > I have seen code, do not find it anymore in code base, which use cmpxchg for Atomic::release_store_fence(). (which is "new");
> 
> ```
> Atomic::store(_x, true);
> Atomic::store_release_fence(_one_way_barrier, true);
> bool z = Atomic::load(_z);
> ```
> 
> But it is written as:
> 
> ```
> Atomic::store(_x, true);
> Atomic::cmpxchg(_one_way_barrier, false, true);
> bool z = Atomic::load(_z);
> ```
> 
> Where load Z happens after store X. But without release the store may happens after.

Yuck. I hope that is never going to return.

I think all such usages of cmpxchg were replaced by an atomic add of zero into the stack, which works on x86. 

> As we say: "All of the atomic operations that imply a read-modify-write action guarantee a two-way memory barrier across that operation."

Well, that's for the HotSpot internal atomics, rather than for Java code. And given that this particular version of `cmpxchg` is only used in a few places it suffices to see what they need. 

> I have look over the cmpxchg uses and not releasing 'seems' okay at first glance. But I think before this line changes we should release on failed CAS.

I have never agreed with that thinking. We should find the callers of CAS that depend on such a side effect and fix them. The problem is that no-one wants to do it. There may no longer be any such cases.

**Review by RealFYang — COMMENTED** [review]
Two minor comments remain. What about the test coverage?

**Comment by luhenry** [comment]
> What about the test coverage?

it's been tested on QEMU running hotspot tier1, everything is passing.

**Review by RealFYang — APPROVED** [review]
Thanks for the update. Seems fine given that it is an experimental feature for now. We will need more thorough tests (jcstress, benchmark, etc.) on real hardware when turning this into a product option.

**Review by robehn — APPROVED** [review]
Thanks

**Comment by luhenry** [comment]
> Thanks for the update. Seems fine given that it is an experimental feature for now. We will need more thorough tests (jcstress, benchmark, etc.) on real hardware when turning this into a product option.

I'm sure we'll be the first one to thoroughly test it and make sure everything works as expected! Thanks
